reasoner/planner/CostEstimator.java

```diff
@@ -0,0 +1,384 @@
+    }
+
+    public long estimateAnswers(ResolvableConjunction conjunction, Set<Variable> variablesOfInterest, Set<Resolvable<?>> includedResolvables) {
+        registerConjunction(conjunction);
+        return costEstimators.get(conjunction).estimateAnswers(variablesOfInterest, includedResolvables);
+    }
+
+    private void registerConjunction(ResolvableConjunction conjunction) {
+        if (!costEstimators.containsKey(conjunction)) {
+            costEstimators.put(conjunction, new ConjunctionAnswerCountEstimator(this, conjunction));
+            costEstimators.get(conjunction).initializeOnce();
+        }
+    }
+
+    private static class ConjunctionAnswerCountEstimator {
+
+        private final ResolvableConjunction conjunction;
+        private final CostEstimator costEstimator;
+        private final LogicManager logicMgr;
+        private Map<Variable, CostCover> unaryCostCover;
+        private Map<Resolvable<?>, List<LocalEstimate>> multivarEstimates;
+        private long fullAnswerCount;
+
+        public ConjunctionAnswerCountEstimator(CostEstimator costEstimator, ResolvableConjunction conjunction) {
+            this.costEstimator = costEstimator;
+            this.logicMgr = costEstimator.logicMgr;
+            this.conjunction = conjunction;
+            this.fullAnswerCount = -1;
+        }
+
+        public long estimateAllAnswers() {
+            return this.fullAnswerCount;
+        }
+
+        public long estimateAnswers(Set<Variable> projectToVariables, Set<Resolvable<?>> includedResolvables) {
+            List<LocalEstimate> enabledEstimates = iterate(includedResolvables)
+                    .flatMap(resolvable -> iterate(multivarEstimates.get(resolvable)))
+                    .toList();
+            return computeCostCover(enabledEstimates, projectToVariables);
+        }
+
+        private long computeCostCover(List<LocalEstimate> multivarEstimates, Set<Variable> projectToVariables) {
+            // Does a greedy set cover
+            Map<Variable, CostCover> costCover = new HashMap<>(unaryCostCover);
+            multivarEstimates.sort(Comparator.comparing(x -> x.answerEstimate(this, projectToVariables)));
+            for (LocalEstimate multivarEstimate : multivarEstimates) {
+                Set<Variable> interestingSubsetOfVariables = multivarEstimate.variables.stream()
+                        .filter(projectToVariables::contains).collect(Collectors.toSet());
+
+                long currentCostToCover = CostCover.costToCover(interestingSubsetOfVariables, costCover);
+                if (currentCostToCover > multivarEstimate.answerEstimate(this, interestingSubsetOfVariables)) {
+                    CostCover variablesNowCoveredBy = new CostCover(multivarEstimate.answerEstimate(this, interestingSubsetOfVariables));
+                    interestingSubsetOfVariables.forEach(v -> costCover.put(v, variablesNowCoveredBy));
+                }
```

Comment:
Do a greedy cover for the set of variables being queried, based on constraints applied by the resolvables included.